Mapbox gl with clustering

Chrissy opened the pull request:
![video_play](https://cloud.githubusercontent.com/assets/435981/24643864/f568b128-18d5-11e7-99c2-1d2b0e19be88.gif)

stuff left to do:

- [x] get mapboxgl to play nicely with webpack
- [x] figure out how to add color to the pin icons (surprisingly hard!)
- [x] move all of these big "mapbox style" objects into a json file somewhere 

I would like to point out that this pull uses not one BUT TWO features from the latest mapbox-gl changelog:
![image](https://cloud.githubusercontent.com/assets/435981/24671350/d872d1dc-1936-11e7-9d35-6bceb8269f51.png)

stuartsan commented:
[just peeked at this locally, it is looking dope!]

Chrissy commented:
@stuartsan nice! glad it worked lol

Chrissy commented on src/assets/styles/index.scss:
```diff
@@ -7,7 +7,8 @@
 @import "~styles/event_card";
-@import "~styles/pikaday";
+@import "~styles/vendor/pikaday";
+@import "~styles/vendor/mapbox-gl.css";
```
this file is located at `node_modules/mapbox-gl/dist/mapbox-gl.css`. It didn't seem like a good idea to import stuff directly from the module itself, so I moved it here, but I'm also open to other things as this seems a little weird too.

stuartsan replied:
IMO importing from node_modules is probably ok.....tim is doing that over here for the date picker as well, to import base pikaday styles then override: https://github.com/aclu-people-power/map/pull/69/files#diff-aeb7f7b7447880522fbc3ecbe64310daR1

We could probably do something fancier with the webpack config too but IMO it would probably be about the same end result :man_shrugging:

Chrissy commented on src/components/EventMap.js:
```diff
@@ -1,9 +1,12 @@
 import Vue from 'vue';
+import { getFilteredEvents } from 'src/util/events';
+import mapMarker from 'src/assets/images/map_marker.png';
+import mapStyles from 'src/assets/styles/mapbox_styles';
```
I've never really been sure where to put styles in the [mapbox specification](https://www.mapbox.com/mapbox-gl-js/style-spec/). I put them in with the other `scss` styles but again, I'm open to other ideas.

stuartsan commented on package.json:
```diff
@@ -42,6 +42,8 @@
     "postcss-loader": "^1.3.3",
     "sass-loader": "^6.0.3",
     "style-loader": "^0.16.0",
+    "turf-distance": "^3.0.12",
```
I think the package.json needs to drop mapbox.js and add mapbox-gl...we recently switched to yarn instead of npm for package management, so the simplest way to achieve that is, after installing yarn, `yarn add mapbox-gl` and `yarn remove mapbox.js`...besides the package.json, it will make a bunch of changes to the yarn.lock file.

Chrissy replied:
oh right! I'm new to Yarn but hear great things. I'll install and update.

stuartsan commented on src/util/events.js:
```diff
@@ -57,10 +58,8 @@ export function computeFilteredEvents(events, filters, zipcodes) {
-
       const milesFromZipcode = metersToMiles(
-         zipcodeLatLng.distanceTo([event.lat, event.lng])
+        distance([zipcodes[filters.zipcode][1], zipcodes[filters.zipcode][0]], [event.lng, event.lat])
```
Note: we are generating the valid zipcodes file, we could reorg the coords order to be `[lng, lat]` if that is more useful/in line with mapbox-gl , seems like this would make a couple things in here slightly simpler...LMK if you think it is worth it, happy to make that change!

Chrissy replied:
it seems worth it to me, tho maybe better/easier to change it on both ends in a follow-up pull?

stuartsan replied:
Yeah totally, let's address that separtely

stuartsan commented on src/components/EventMap.js:
```diff
@@ -61,50 +65,60 @@ export default function(store){
+      addCustomIcon(icon, name) {
+        const img = new Image();
+        img.src = icon;
+        img.onload = function() {
```
Nit: can drop the need for bind(this) with a fat arrow function, ` img.onload = () => {`

Chrissy replied:
oh right good call!

stuartsan commented on src/components/EventMap.js:
```diff
@@ -61,50 +65,60 @@ export default function(store){
+      });
+
+      this.mapRef.addControl(new mapboxgl.NavigationControl());
+      this.mapRef.on("load", this.mapMounted.bind(this));
```
More bind nits -- I think vue has already bound this method, should be ok to drop `.bind(this)`

stuartsan commented on src/components/EventMap.js:
```diff
@@ -61,50 +65,60 @@ export default function(store){
-            weight: 2
-          }).addTo(this.eventsLayer);
-        });
+        if (eventsSource) {
```
I feel like might be worth a comment to note when eventsSource does _not_ exist -- I'm not positive but it seems like this could happen if the events data loaded before the map finished loading and `mapMounted` is called

Chrissy replied:
that is a definite possibility! tho in theory if that happened the function would do nothing and then fire again via `mapMounted`. Should the comment just sorta break down what happens if `eventsSource` is false?

stuartsan replied:
Yeah exactly, I was at first unsure if the map would handle this kind of race condition and it totally does, but might be worth a comment here to make it easy to understand at a glance.

stuartsan commented on src/components/EventMap.js:
```diff
@@ -14,6 +16,8 @@ export default function(store){
       mapRef: null,
       eventsLayer: null,
+      initialCoordinates: [-96.9 , 37.8],
+      initialZoom: 4
```
One thing I noticed, my initial zoom level / coordinates are kind of funky, not sure if this is partially cause of my screen size or something, or if the coordinates/zoom need adjustment:

![image](https://cloud.githubusercontent.com/assets/1724544/24691856/da277808-198a-11e7-94b9-79ee2ea7020d.png)

Chrissy replied:
oh woah... @stuartsan are your pin icons still that big if you update this branch? That is NOT the intended size lol.

Chrissy replied:
also: can you give me a few more details about how/when the viewbox feels off? 

* Is the mis-centering happening only when the sidebar is open?
* Is the zoom level too zoomed-in or too zoomed-out?

stuartsan replied:
Oh! Ha ok cool. Yes, in chrome, I do still see the oversized markers on the up to date branch. Zoom level is slightly too zoomed in (plus the miscentering). And I see the same issues with sidebar both open and closed.

I wonder why we see different things 🤔

Chrissy replied:
ok cool! looking into it now.

stuartsan commented on src/assets/styles/mapbox_styles.json:
```diff
@@ -0,0 +1,39 @@
+[{
```
Still tryna wrap my mind around mapbox gl but this stuff is super neat

Chrissy replied:
It's pretty cool but also a bit magic (in a bad way) sometimes.

Depending on how often the data sources update, down the road we could take this a step further and upload the data to mapbox using the [js sdk](https://github.com/mapbox/mapbox-sdk-js) and style the whole thing using the free online editor. that would move all of this stuff out of the codebase. Plus the data would come in with the vector tiles which is pretty sweet.

stuartsan replied:
Oh huh, that is cool!

stuartsan commented:
> I would like to point out that this pull uses not one BUT TWO features from the latest mapbox-gl changelog

🤘 🤘 🤘 

I left some comments, this is all looking legit, thanks very much!! It is cool to see how mapbox gl works vs mapbox.js. I think generally merging this work sooner than later will be good so we can split out some other details and build on top of it.

stuartsan commented:
Nice, these updates seem to have fixed just about all the visual discrepancies I was seeing! There is still one thing I see, only when the event list sidebar is visible: things are centered but zoomed in one click too much:

![image](https://cloud.githubusercontent.com/assets/1724544/24766584/bb670754-1ab0-11e7-842b-3bbacf0da8cb.png)

But if you type in a zipcode, and then remove it, it correctly fits the US bounds.

Poking at it a teeny bit I noticed that moving the call to `map.resize()` into the top of `setMapPositionBasedOnZip()` fixes this; not sure why but it seems like that might make sense generally, ensuring the resize always happens right before setting the map "position"...

Chrissy commented:
@stuartsan yeah I was actually worried about a bug like that. Mapbox only fires a `resize` event when the window resizes, not when a containing element resizes, so dom changes won't trigger anything on the mapbox end.

Is there a wholesale `render()` type method to hook into for something like that? `plotEvents` would work fine since it fires pretty frequently, though it feels a bit incidental...

stuartsan commented:
Ahh yes, gotchya. There is an `updated()` hook:

https://vuejs.org/v2/api/#updated

And I think this would be the right place to do that probably, guessing something like

```js
updated() {
  if (this.mapRef) {
    this.mapRef.resize();
  }
}
```

Chrissy commented:
@stuartsan ok i think i fixed it, if not _exactly_ in the way we intended. `updated()` wasn't firing, either because it only fires on updates (not prop inits) or because changes from a data store don't trigger that function. ANYWAYS

there are really only two scenarios:

1. event list arrives, dom updates, map mounts
2. map mounts, event list arrives, dom updates

I corrected for the second scenario by adding a `reset()` call inside the `events()` watcher. It worked on 10/10 refreshes including slow connections on a virtual machine so hopefully we are solid!

Unless you see something else I think this thing is ready to go! Lemme know what the process is for merge!

stuartsan commented:
@Chrissy , thanks for digging into that! Sounds good to me. I'll go ahead and merge this -- not much of a process besides doing "merge pull request" via the GH UI. Thanks!

Chrissy commented:
🎉 🎈 🎊
